core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -1753,7 +1756,7 @@ private <C extends Comparable<C>> RexNode simplifyAnd2ForUnknownAsFalse(
         if (literal2 == null) {
           continue;
         }
-        if (!literal1.equals(literal2)) {
+        if (literal1.getType().equals(literal2.getType()) && !literal1.equals(literal2)) {
```

Comment:
I'm puzzled about the necessity of adding extra type equals checks. What specific issues might arise without them in SQL simplification?